cpp/src/io/comp/gpuinflate.cu

```diff
@@ -0,0 +1,1291 @@
+    int16_t offs[MAXBITS + 1];  // offset in symbol table for each length (scratch)
+};
+
+struct lut_arr
+{
+    int32_t lenlut[1 << LOG2LENLUT];    // LUT for length decoding
+    int32_t distlut[1 << LOG2DISTLUT];  // LUT for fast distance decoding
+};
+
+
+// 4 batches of 32 symbols
+#define LOG2_BATCH_COUNT   2 // 1..5
+#define LOG2_BATCH_SIZE    5
+#define BATCH_COUNT        (1 << LOG2_BATCH_COUNT)
+#define BATCH_SIZE         (1 << LOG2_BATCH_SIZE)
+
+struct xwarp_s
+{
+    int32_t batch_len[BATCH_COUNT]; // Length of each batch - <0:end, 0:not ready, >0:symbol count
+    union {
+        uint32_t symqueue[BATCH_COUNT * BATCH_SIZE];
+        uint8_t symqueue8[BATCH_COUNT * BATCH_SIZE * 4];
+    } u;
+};
+
+#define ENABLE_PREFETCH     1
+
+#if ENABLE_PREFETCH
+#define LOG2_PREFETCH_SIZE  9 // Must be at least LOG2_BATCH_SIZE+3
+#define PREFETCH_SIZE       (1 << LOG2_PREFETCH_SIZE)
+
+#define PREFETCH_ADDR32(q, p)    (uint32_t *)(&q.pref_data[(PREFETCH_SIZE - 4) & (size_t)(p)])
+struct prefetch_queue_s
+{
+    const uint8_t *cur_p; // Prefetch location
+    int run; // prefetcher will exit when run=0
+    uint8_t pref_data[PREFETCH_SIZE];
+};
+
+#endif // ENABLE_PREFETCH
+
+
+struct inflate_state_s {
+    // output state
+    uint8_t *out;           // output buffer
+    uint8_t *outbase;       // start of output buffer
+    uint8_t *outend;        // end of output buffer
+    // Input state
+    uint8_t *cur;           // input buffer
+    uint8_t *end;           // end of input buffer
+
+    uint2    bitbuf;        // bit buffer (64-bit)
+    uint32_t bitpos;        // position in bit buffer
+
+    int32_t err;
+    int     btype;          // current block type
+    int     blast;          // last block
+    uint32_t stored_blk_len;// length of stored (uncompressed) block
+
+    uint16_t first_slow_len;    // first code not in fast LUT
+    uint16_t index_slow_len;
+    uint16_t first_slow_dist;
+    uint16_t index_slow_dist;
+
+    volatile xwarp_s x;
+#if ENABLE_PREFETCH
+    volatile prefetch_queue_s pref;
+#endif
+
+    int16_t lencnt[MAXBITS+1];
+    int16_t lensym[FIXLCODES]; // Assumes FIXLCODES >= MAXLCODES
+    int16_t distcnt[MAXBITS+1];
+    int16_t distsym[MAXDCODES];
+
+    union {
+        scratch_arr scratch;
+        lut_arr lut;
+    } u;
+};
+
+inline __device__ unsigned int bfe(unsigned int source, unsigned int bit_start, unsigned int num_bits)
+{
+    unsigned int bits;
+    asm("bfe.u32 %0, %1, %2, %3;" : "=r"(bits) : "r"(source), "r"(bit_start), "r"(num_bits));
+    return bits;
+};
+
+
+inline __device__ uint32_t showbits(inflate_state_s *s, uint32_t n)
+{
+    uint32_t next32 = __funnelshift_rc(s->bitbuf.x, s->bitbuf.y, s->bitpos);
+    return (next32 & ((1 << n) - 1));
+}
+
+inline __device__ uint32_t nextbits32(inflate_state_s *s)
+{
+    return __funnelshift_rc(s->bitbuf.x, s->bitbuf.y, s->bitpos);
+}
+
+inline __device__ void skipbits(inflate_state_s *s, uint32_t n)
```

Comment:
There seems to be at least some code duplication between this file and debrotli.cu -- possible to consolidate? (I haven't done a close comparison.)

Reply:
I would rather keep each codec fully independent from each other (IMO this is a case dependency hygiene trumps code reuse). The amount of code that can be shared is relatively small and could change depending on codec-specific optimizations (decompression testing, comparing with reference CPU version has been done primarily outside of cudf, since cudf doesn't include the reference implementation).